### ISSUE TITLE
[Docs] Add rmlUI docs

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/RecoilEngine#1672]
Examples, API listings etc.

AFAICT Lua interface seems to be defined in these functions, dunno about the rest:
https://github.com/beyond-all-reason/spring/blob/432f997594b38f4232cf5e7d9f79ba8acd97e57f/rts/Rml/SolLua/RmlSolLua.cpp#L71-L81

**Comment by lhog** [comment]
btw, we need to rename these functions. The naming style looks completely alien compare to the rest of the Lua API.

**Comment by loveridge** [comment]
The lua api mostly follows the official implementation, which has documentation here: https://mikke89.github.io/RmlUiDoc/pages/lua_manual/api_reference.html 

We will need an article for things that deviate from this. I recall a few cases:

- `context:LoadDocument(filepath, widget)` takes a second argument that is the widget itself.
- Inline event handlers can call widget functions like so: `onclick="widget:OnReloadClick()"`. `document` is available too?
- Documentation for any custom components. I think just Chris' Texture one at the moment.
- Our data model is reactive only on the first depth level.
- Does data model need to have keys assigned at creation for reactivity? I forget

And then general guidance for how RmlUi specifically integrates with Recoil widgets.

**Comment by ChrisFloofyKitsune** [comment]
Yup, takes an extra parameter to bind to the "widget" parameter in the document's Lua context

That's so inline event handlers can happen.

The only custom component is the texture one. I've not found any reason to make any more than that.

Is the data model in the default RmlUi Lua implementation reactive on more than just the first level?
I can look into changing that?

The way I've set up the data model all keys for reactivity are expected to exist at creation. I'm not quite sure what happens/should happen if the same data model is "opened" repeatedly.

The big RmlUi examples widget I've been working on is intended to act as a library of in-game examples with accompanying code

**Comment by loveridge** [comment]
I think reactivity at any depth on the data model would be ideal.

**Comment by sprunk** [comment]
#1955 adds reactivity. I don't think the lack of reactivity was documented anywhere so this is just an unactionable remark.

**Comment by badosu** [comment]
I've [sent a comment](https://github.com/mikke89/RmlUi/discussions/186#discussioncomment-12917084) on a resources related discussion about generated lua documentation. In the hope of either them having a pointer on how to generate the docs ourselves or grab from somewhere if they exist already.

**Comment by badosu** [comment]
> We will need an article for things that deviate from this. I recall a few cases:

According to [RmlSolLua readme](https://github.com/LoneBoco/RmlSolLua?tab=readme-ov-file#coverage), they add a considerable amount of new bindings that, as far as I can tell from reading the linked section, the only way to document is looking at the source.

**Comment by badosu** [comment]
@ChrisFloofyKitsune @loveridge Just so we can understand the scope of rmlui Recoil Lua API documentation effort:

- Is any global exposed to the lua environment except for `RmlUi`?
- Is is exposed only in `LuaUI/main.lua` or any other environment such as `LuaRules/draw.lua` (unsynced luarules)?
- Assuming we import a subset (please correct me if I'm wrong) of the full RmlUI Lua API, should it be simple to figure out what to pull documentation from looking at a file (e.g. the source sprung shared above)?
- Some divergence from the "vanilla" API, not listed in this issue (texture, maybe something else?), would be good to be known for the person who fixes this issue.
- I understand we use `RmlSolLua` for the bindings right? It's safe to assume we pull their extended API as well (I remarked on that above)?

**Comment by loveridge** [comment]
- Yes, all of the rmlui lua functions are exposed through `RmlUi` global.
- RmlUi lua functions can only be accessed from unsynced luaui, no luarules.
- All of the lua functions listed in the official docs should be supported, with the exception of 5.0 apis I suppose. The initial rml lua integration work was completed before 5.0 was released, and I don't know what is new. Sprung is correct that those `bind_` functions contain the sol bindings to the C++ functions. For example https://github.com/beyond-all-reason/RecoilEngine/blob/master/rts/Rml/SolLua/bind/Element.cpp#L196
- The divergences I recall are listed in https://github.com/beyond-all-reason/RecoilEngine/issues/1672#issuecomment-2327084316 Since that comment, I added `__SetDirty` to the data model object that sets a particular property of the data model to be dirty.
- RmlSolLua was indeed the origin of our bindings
